rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+enabled. I had hoped to provide a dependency graph of these features but their
+number and relation meant that what was generated was not explicable. Some
+features seem to do almost but not exactly quite the same things.
+
+Taken in the abstract the core of vector is a data ingest, transformation and
+egress framework, expressed as an acyclic graph with nodes being separated by a
+queue -- both in-memory and on-disk -- with associated mechanisms providing
+durability across restarts, acks, back-pressure between nodes in the graph and
+self-instrumentation. Nodes are of type "source", "sink" and "transform". A
+"source" node creates new `Event` instances -- this is vector's internal data
+type -- and a "sink" destroys them, possibly by egressing them but also possibly
+by just deallocating. A "transform" node modifies `Event` instances as they pass
+through, destroys them, merges them or creates additional `Event`
+instances. Configuration and reload management of, implementations of
+source/transform/sink for different domains, common mechanism for backoff/retry
+in sinks and cross-cutting concerns like tracing integration are all non-core.
+
+Our project today has a small number of packages in `lib`, notably:
+
+* `lib/shared` -- a collection of common-ish types and functions, vector's "util"
+* `lib/file-source` -- the backing code for the file source, minus some of, but
+  not all of, the hook-up to be a proper vector source
+* `lib/vrl` -- the implementation of VRL in its entirety
+
+VRL is the gold standard in vector as of this writing. It's a package containing
+several sub-crates with a gentle dependency graph. From the point of view of
+vector there is a single interface to cope with but VRL developers have
+_additional_, unexposed interfaces to aid their work and experimentation.
+
+The top-level package in vector is where most of the code lives. This code is
+relied on by its sub-packages and, so, contains pretty well every major concern
+of vector, all sources, sinks and transforms. There are some self-contained
+modules here -- `src/metrics` for example -- but someone working on these
+modules will have to participate in full vector build times to test their
+changes. Previous discussions concerning core extraction have noted that there
+are clear division lines in the top-level. "All AWS related code could live in a
+package, under a feature flag," which is true. All AWS related code could live
+in the top-level crate but under a feature flag, equally well.
+
+Our present feature flag crowd makes our testing surface high, though our CI
+tooling has done a good job of catching any features I have forgot to enable in
+addition to those present in "default" feature when I make changes. Tools like
+[cargo-hack](https://github.com/taiki-e/cargo-hack) or
+[cargo-all-features](https://lib.rs/crates/cargo-all-features) will help as
+well.
+
+## What is "core" to vector?
+
+I argue that the following are "core":
+
+* the `Event` type,
```

Comment:
Probably implicit by pulling in `Event`, but the `Value` type is pretty essential too.

Reply:
I have added `Value`